### PR TITLE
Squashing issues with figure tests

### LINKED ISSUE #1861: tests create figure hashes directory
when running `py.tests` on single file it creates unneeded tmp directories.

### LINKED ISSUE #1873: Port the figure tests to Python 3
Using Python 3.5 and trying to write a new figure hash table, I got this error.  Googling the TypeError at the end shows that this is a Python 2/3 issue.

http://stackoverflow.com/questions/33054527/python-3-5-typeerror-a-bytes-like-object-is-required-not-str

```
All test files for figure hashes can be found in /tmp/tmp_xhdqdu9_figures
Traceback (most recent call last):
  File "/home/ireland/anaconda3/bin/py.test", line 6, in <module>
    sys.exit(py.test.main())
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/config.py", line 49, in main
    return config.hook.pytest_cmdline_main(config=config)
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
    res = hook_impl.function(*args)
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/main.py", line 119, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/main.py", line 115, in wrap_session
    config._ensure_unconfigure()
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/config.py", line 844, in _ensure_unconfigure
    self.hook.pytest_unconfigure(config=self)
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/home/ireland/anaconda3/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
    res = hook_impl.function(*args)
  File "/home/ireland/projects/sunpy/sunpy/sunpy/conftest.py", line 70, in pytest_unconfigure
    json.dump(hash.hash_library, outfile, sort_keys=True, indent=4, separators=(',', ': '))
  File "/home/ireland/anaconda3/lib/python3.5/json/__init__.py", line 179, in dump
    fp.write(chunk)
TypeError: a bytes-like object is required, not 'str'
```

### PR DESCRIPTION
**Pull request by ayshih**
1) Fixes #1873 
Figure-hash tables can now be created on Python 3.5, and I've implemented separate hash tables for different Python major versions.

2) Fixes #1861 
A temporary directory for figure-test images is created only if at least one figure-test image was created

3) Fixes bug where figure-test images were not preserved if the corresponding figure test failed (mentioned in #1965)

4) The hash library is now always written out, and placed in the same directory as the PNG images
Previously, a hash library was written out only if hashes needed to be added (either because a new test was created or because the previous JSON file was explicitly deleted).  My intent was to avoid creating a new library if one was not needed (see https://github.com/sunpy/sunpy/pull/1277#r23945809).  However, since PNG images are now always saved (#1631), we may as well do the same with the JSON library.  Thus, the workflow for updating a hash library is more straightforward: just pluck it out of the temp directory and overwrite the existing library under `sunpy/tests/`.

5) Travis CI now runs the figure tests under Python 3.5 in addition to Python 2.7, but because we haven't yet formally defined a complete test environment for Python 3.5, the tests are allowed to fail

### CODE REVIEW

**Comment by ayshih** [comment]
By the way, the four figure hashes that don't agree between Python 2.7 and Python 3.5 appears to be the result of a slight difference in the vertical position of the X-axis label when not using WCSAxes.  I guess that's one more reason to use WCSAxes! =P

**Comment by Cadair** [comment]
Thanks @ayshih could you change the Travis build to use 3.5 for the figure tests in this PR as well just for good measure?

**Comment by Cadair** [comment]
good @pep8speaks bot, you checked for line length to 100 char. :)

File: .travis.yml [hunk 1]
```diff
@@ -52,7 +52,10 @@ matrix:
            env: JOB="Documentation" PYTHON_VERSION=3.5 SETUP_CMD='build_sphinx -w'
 
          - os: linux
-           env: JOB="Figures" PYTHON_VERSION=2.7 SETUP_CMD='test --figure' CONDA_DEPENDENCIES=''
+           env: JOB="Figures 2.7" PYTHON_VERSION=2.7 SETUP_CMD='test --figure' CONDA_DEPENDENCIES=''
```
Comment (Cadair) [inline .travis.yml]: Do we want to run both on the CI? They take a fair while to run.

Comment (ayshih) [inline .travis.yml]: They only take ~5 minutes each, and it makes sure that people update both hash tables when an update is needed.

If time is a real issue, we could instead have `python setup.py test --figure` test *only* figures.

Comment (Cadair) [inline .travis.yml]: ok, I thought it took longer than that +1 to both.

**Comment by ayshih** [comment]
@Cadair Regarding the merge conflict in `.travis.yml`, I'm fuzzy on the configuration particulars.  Should I keep the `CONDA_DEPENDENCIES=''` and is the comment correct?  (I see your commits aef24d2 and e568f94.)

**Comment by Cadair** [comment]
@ayshih Haha, yes, the `CONDA_DEPENDENCIES=''` flag should be kept as it prevents the whole environment being installed twice. for the figure tests we download the environment yaml from the sunpy/figure-tests repo: https://github.com/ayshih/sunpy/blob/9f5b7d29995397790d7d3817347b826a10ac66b8/.travis.yml#L82

We will need a new version of that yml file for Python 3.5 and have to write some selective download logic in bash oneliners :s

**Comment by ayshih** [comment]
Ah, okay, now I'm understanding.  That's why the "Figures 3.5" that I created for Travis doesn't actually run the figure tests under Python 3.5 because the line that you referenced creates *another* environment, which forces Python 2.7.

**Comment by ayshih** [comment]
This PR is ready to merge, but the figure tests on Python 3.5 are currently allowed to fail on Travis because they only succeed on the previous version of Astropy (1.2.1).  We'll need to either create a 3.5-specific environment file or merge #1965 (or both)

**Comment by Cadair** [comment]
I have restarted the online build but any failures are not related to this PR.

**Comment by ayshih** [comment]
Okay, let's try this again.  This PR is ready to merge, with the following notes:

* Versions of matplotlib from conda-forge apparently give slightly different label placement than versions from the primary channel.  So, reproducing the hashes in the 3.5 hash library requires a conda-forge build (<2.0.0, because that version completely changes things), and reproducing the hashes in the 2.7 hash library may require a non-conda-forge build.  Just roll with it.
* Astropy 1.3 doesn't appear to give the right plots without #1965.

**Pull request merged**